### PR TITLE
[Program Capture] Capture & execute `qml.grad` in plxpr

### PR DESCRIPTION
**Pull request by dwierichs**
**Context:**
The new `qml.capture` module does not support differentiation yet.

**Description of the Change:**
This PR takes the first step towards differentiability in plxpr.
It adds the capability of capturing `qml.grad` as a "nested jaxpr" primitive.
When executing the captured program, `qml.grad` is essentially changed to `jax.grad`, because executing Autograd autodifferentiation within the Jaxpr ecosystem is not sensible.

**Benefits:**
Capture first differentiation instructions

**Possible Drawbacks:**
The current implementation requires a `jvp` construction for every evaluation of a QNode gradient. This means that this JVP function is reconstructed for every evaluation call, if I'm not mistaken, making the code significantly less performant with `capture` than without. Of course, the longer term plan is to process the plxpr into lower-level code by lowering the `grad` primitive itself, in which case this problem goes away.
A similar redundancy is implemented in `QNode`: Whenever a `qnode` primitive is evaluated, a new `QNode` is created (and only ever evaluated once). This disables caching, for example, unless a cache is passed around explicitly.

**Related GitHub Issues:**

[sc-71858]

### CODE REVIEW

File: pennylane/_grad.py [hunk 1]
```diff
@@ -15,19 +15,39 @@
...
 from autograd.wrap_util import unary_to_nary
 
+from pennylane.capture import create_grad_primitive, enabled
 from pennylane.compiler import compiler
 from pennylane.compiler.compiler import CompileError
 
 make_vjp = unary_to_nary(_make_vjp)
 
 
+def _capture_diff(func, argnum=None, diff_prim=None):
```
Comment (dwierichs) [inline pennylane/_grad.py:33]: This will be used for `qml.grad` and `qml.jacobian`. The kwarg `diff_prim` specifies which one it is.

File: pennylane/_grad.py [hunk 2]
```diff
@@ -15,19 +15,39 @@
...
+    """Capture-compatible gradient computation."""
+    import jax  # pylint: disable=import-outside-toplevel
+
+    if isinstance(argnum, int):
+        argnum = [argnum]
+    if argnum is None:
+        argnum = [0]
```
Comment (dwierichs) [inline pennylane/_grad.py:40]: As we are dispatching to `jax.grad` and `jax.jacobian` for captured `qml.grad` (`qml.jacobian`), we reproduce the default argnum in those JAX functions.

File: pennylane/capture/capture_qnode.py [hunk 3]
```diff
@@ -87,6 +89,11 @@ def _(*args, qnode, shots, device, qnode_kwargs, qfunc_jaxpr, n_consts):
         mps = qfunc_jaxpr.outvars
         return _get_shapes_for(*mps, shots=shots, num_device_wires=len(device.wires))
 
+    def _qnode_jvp(*args_and_tangents, **impl_kwargs):
+        return jax.jvp(partial(qnode_prim.impl, **impl_kwargs), *args_and_tangents)
```
Comment (dwierichs) [inline pennylane/capture/capture_qnode.py:93]: The QNode primitive receives its own `JVP` rule here in order to be differentiable. It is simply a call to `jvp` acting on the `impl` call of the primitive. This dispatches to the standard derivative registered for QNodes in PennyLane with jax.

File: pennylane/capture/primitives.py [hunk 4]
```diff
@@ -163,6 +163,31 @@ def __hash__(self):
...
+
+        def bind_with_trace(self, trace, args, params):
+            """Bind the ``NonJVPPrimitive`` with a trace. If the trace is a ``JVPTrace``,
+            binding falls back to a standard Python function call. Otherwise, the
+            bind call of JAX's standard Primitive is used."""
+            if isinstance(trace, jax.interpreters.ad.JVPTrace):
+                return self.impl(*args, **params)
+            return super().bind_with_trace(trace, args, params)
+
+    return NonJVPPrimitive
```
Comment (dwierichs) [inline pennylane/capture/primitives.py:188]: We want this new class of primitive for operators, which is handled like a standard Python function by `JVPTrace`. This is because the differentiation of PennyLane operators does _not_ happen through them returning a modified operator or an "OperatorDerivative" type, but simply by storing the `JVPTracer` in the created operator instance.
This way, the differentiation of the QNode via `jax.jvp` can go its standard path.

**Review by dwierichs — COMMENTED** [review]
Just leaving some (hopefully helpful) comments.

File: pennylane/capture/capture_qnode.py (hunk 3 above)
Comment (dwierichs) [reply pennylane/capture/capture_qnode.py:93]: Yes, because we're dispatching back to the standard computational path, which supports preprocessing. This is also why parameter-shift rule, and in fact any diff method working without capture, works with capture, too. It's just a wrapper at the moment.

**Review by albi3ro — COMMENTED** [review]
Excellent work finding a workaround with the `NonJVPPrimtiive`.  Would we also need to use this on the other higher-order primitives (`adjoint`, `cond`, `for_loop`, etc.)?

So I tried combining this work with PR #6123 . 🤞 we won't be needing `NonJVPPrimitive` long-term once we have a new pipeline for gradients.  That work probably won't turn into anything in the near future, but it's nice to know that the `NonJVPPrimitive` probably won't need to be a permanent fix.

File: pennylane/capture/primitives.py [hunk 5]
```diff
@@ -318,3 +343,32 @@ def _(*args, has_eigvals=False, **_):
         return abstract_type(abstract_eval, n_wires=n_wires, has_eigvals=has_eigvals)
 
     return primitive
+
+
+@lru_cache
+def create_grad_primitive():
```
Comment (albi3ro) [inline pennylane/capture/primitives.py:349]: Make `get` instead of `create`? Since most of the time we're just retrieving one that's already been created.

File: pennylane/_grad.py [hunk 6]
```diff
@@ -97,9 +117,10 @@ def __new__(cls, func, argnum=None, method=None, h=None):
             return ops_loader.grad(func, method=method, h=h, argnums=argnum)
 
         if method or h:  # pragma: no cover
-            raise ValueError(
-                f"Invalid values for 'method={method}' and 'h={h}' in interpreted mode"
-            )
+            raise ValueError(f"Invalid values '{method=}' and '{h=}' without QJIT.")
```
Comment (albi3ro) [inline pennylane/_grad.py:120]: Given we want to be able to have catalyst interpret `qml.grad` captured into jaxpr and use that instead, I think we should be able to capture the catalyst keyword arguments as well.

Any necessary validation can occur be whatever consumes the `qml.grad` primitive.

File: pennylane/capture/primitives.py (hunk 5 above)
Comment (dwierichs) [reply pennylane/capture/primitives.py:349]: Ah, yeah good point!
- I was following the naming of the other functions in this file.
- I suppose within PL itself we only use this function once, and then it is creating the primitive...
That being said, I don't have a strong preference. Should I change?

File: pennylane/_grad.py (hunk 6 above)
Comment (dwierichs) [reply pennylane/_grad.py:123]: Done

**Comment by dwierichs** [comment]
@albi3ro I addressed all comments and implemented higher-order primitives support + tests. Regarding the naming of `create_grad_primitive` I don't have a strong preference.

File: tests/capture/test_capture_diff.py [hunk 7]
```diff
@@ -0,0 +1,253 @@
...
+    assert qml.math.allclose(manual_eval, jax_out)
+
+    jax.config.update("jax_enable_x64", initial_mode)
+
+
+@pytest.mark.parametrize("x64_mode", (True, False))
+def test_nested_grad(x64_mode):
+    """Test that nested qml.grad primitives can be captured.
+    We use the function
+    f(x) = sin(x)^3
```
Comment (albi3ro) [inline tests/capture/test_capture_diff.py:114]: ❤️

**Review by albi3ro — APPROVED** [review]
🚀

**Review by albi3ro — COMMENTED** [review]
Mind making the primitive accessible in `qml.capture.primitives`?

File: pennylane/capture/capture_operators.py [hunk 8]
```diff
@@ -101,7 +103,7 @@ def create_operator_primitive(
     if not has_jax:
         return None
 
-    primitive = jax.core.Primitive(operator_type.__name__)
+    primitive = create_non_jvp_primitive()(operator_type.__name__)
```
Comment (mudit2812) [inline pennylane/capture/capture_operators.py:106]: 😍

File: pennylane/compiler/qjit_api.py [hunk 9]
```diff
@@ -621,7 +622,7 @@ def _get_for_loop_qfunc_prim():
 
     import jax  # pylint: disable=import-outside-toplevel
 
-    for_loop_prim = jax.core.Primitive("for_loop")
+    for_loop_prim = create_non_jvp_primitive()("for_loop")
```
Comment (mudit2812) [inline pennylane/compiler/qjit_api.py:625]: Don't we need this for `while_loop` as well?

File: tests/capture/test_capture_diff.py [hunk 10]
```diff
@@ -0,0 +1,252 @@
...
+    f'(x) = 3 sin(x)^2 cos(x)
+    f''(x) = 6 sin(x) cos(x)^2 - 3 sin(x)^3
+    f'''(x) = 6 cos(x)^3 - 12 sin(x)^2 cos(x) - 9 sin(x)^2 cos(x)
+    """
+    initial_mode = jax.config.jax_enable_x64
```
Comment (mudit2812) [inline tests/capture/test_capture_diff.py:118]: We could probably put the precision toggling stuff inside a fixture. That way the precision can be automatically parametrized and we don't need to reset the precision manually at the end of each test.

**Review by mudit2812 — CHANGES_REQUESTED** [review]
Looks great! The primitive for `qml.while_loop` needs to be replaced with `NonJVPPrimitive` as well, along with tests for it. I also left an optional comment about cleaning up the tests

File: tests/capture/test_capture_diff.py (hunk 10 above)
Comment (mudit2812) [reply tests/capture/test_capture_diff.py:118]: The current implementation is also used throughout the rest of the capture tests, so it's by no means needed. It could be worth just doing that in a separate PR for all tests in the `capture` folder.

Comment (dwierichs) [reply tests/capture/test_capture_diff.py:118]: Good idea! I agree that we should do this clean up in a separate PR, to get these ones in swiftly.

File: pennylane/compiler/qjit_api.py (hunk 9 above)
Comment (dwierichs) [reply pennylane/compiler/qjit_api.py:625]: Good catch, thanks! Done.

**Review by mudit2812 — APPROVED** [review]
🎉
